### PR TITLE
Added OptionalType which contains all the optional value functionality

### PR DESCRIPTION
**Pull request by jvprat**
I've factored out the optional functionality into its own wrapper type. It can be transparently applied to any type, so there's no more need to have checks for optional on every type.

### CODE REVIEW

File: spec/tableshape_spec.moon [hunk 1]
```diff
@@ -314,10 +314,10 @@ describe "tableshape", ->
 
     it "repairs using repair option callback", ->
       int_string = types.pattern "^%d+$", {
-        optional: true
```
Comment (leafo) [inline spec/tableshape_spec.moon]: I'd like to keep this syntax, I'm already using it in a few of my projects

Comment (jvprat) [inline spec/tableshape_spec.moon]: Okay. Would you consider merging it if I re-add this option?

Comment (leafo) [inline spec/tableshape_spec.moon]: Sure, I like the refactor

**Comment by jvprat** [comment]
I've added support for the optional option again but I don't know if this implementation is too intrusive.

**Comment by leafo** [comment]
sorry about the delay, looks fine!

**Pull request merged**